ingest.py

```diff
@@ -139,6 +141,99 @@ def does_vectorstore_exist(persist_directory: str) -> bool:
                 return True
     return False
 
+
+def prompt_user():
+    """
+    This function prompts the user to select an existing directory or create a new one to store source material.
+    If an existing directory is selected, it checks if the directory is empty and prompts the user to create files
+    in the directory if it is empty. It sets the directory paths as environment variables and returns them.
+
+    Returns:
+    selected_directory_path (str): The path of the selected directory.
+    selected_db_path (str): The path of the database directory for the selected directory.
+    """
+
+    def _get_user_choice(timeout):
+        if not isinstance(timeout, int):
+            raise ValueError("Timeout value should be an integer.")
+        print("Select an option or 'q' to quit:")
+        print("1. Select an existing directory")
+        print("2. Create a new directory")
+        print(f"3. Use current source_directory: {source_directory}")
+        inputs = [sys.stdin]
+        readable, _, _ = select.select(inputs, [], [], timeout)
+        if readable:
+            return sys.stdin.readline().strip()
+        return "3"
+
+    def _display_directories():
```

Comment:
Nice way of displaying sub-directories in 4 columns if you have a lot of them, also I colored print so it's more visible

```python
import textwrap
import math
```

```python
def _display_directories():
        """
        This function displays the list of existing directories in the ./source_documents directory.
        """
        print(f"\n\033[94mExisting directories in ./source_documents:\033[0m")
        directories = sorted((f for f in os.listdir("./source_documents") if not f.startswith(".")), key=str.lower)

        # Set the desired column width and the number of columns
        column_width = 30
        num_columns = 4

        # Calculate the number of rows needed based on the number of directories
        num_rows = math.ceil(len(directories) / num_columns)

        # Print directories in multiple columns
        for row in range(num_rows):
            for column in range(num_columns):
                # Calculate the index of the directory based on the current row and column
                index = row + column * num_rows

                if index < len(directories):
                    directory = directories[index]
                    wrapped_directory = textwrap.shorten(directory, width=column_width - 1, placeholder="...")
                    print(f"{index+1:2d}. {wrapped_directory:{column_width}}", end="")

            print()  # Print a new line after each row

        return directories
```

Reply:
The colouring is a nice touch, but I think for now the columns are a bit too much. When testing this in a small terminal, or resizing it, the columns aren't nicely spaced leading to issues with readability. I think handling all those cases might increase the overall complexity of the interface, at which point it might be worthwhile investing time in a nice GUI instead, what do you think?

Reply:
GUI is definitely a better idea, but then on GUI, you should consider manipulating the change of the dbs sub-directory you want to ask questions to. Also, what if you want to combine them to have joint answers from multiple databases? Maybe something like this but I am not sure. This would be really nice feature to combine topics (folder of books) from the UI:

```python
db = Chroma(persist_directory=chroma_directory, embedding_function=embedding)

db.add_documents(documents=texts1)
db.add_documents(documents=texts2)
```